ext/pdo_firebird/firebird_driver.c

```diff
@@ -756,51 +756,56 @@ static zend_string* firebird_handle_quoter(pdo_dbh_t *dbh, const zend_string *un
 /* }}} */
 
 /* php_firebird_begin_transaction */
-static bool php_firebird_begin_transaction(pdo_dbh_t *dbh) /* {{{ */
+static bool php_firebird_begin_transaction(pdo_dbh_t *dbh, bool is_auto_commit_txn) /* {{{ */
 {
 	pdo_firebird_db_handle *H = (pdo_firebird_db_handle *)dbh->driver_data;
-	char tpb[8] = { isc_tpb_version3 }, *ptpb = tpb+1;
-#ifdef abies_0
-	if (dbh->transaction_flags & PDO_TRANS_ISOLATION_LEVEL) {
-		if (dbh->transaction_flags & PDO_TRANS_READ_UNCOMMITTED) {
-			/* this is a poor fit, but it's all we have */
-			*ptpb++ = isc_tpb_read_committed;
-			*ptpb++ = isc_tpb_rec_version;
-			dbh->transaction_flags &= ~(PDO_TRANS_ISOLATION_LEVEL^PDO_TRANS_READ_UNCOMMITTED);
-		} else if (dbh->transaction_flags & PDO_TRANS_READ_COMMITTED) {
-			*ptpb++ = isc_tpb_read_committed;
-			*ptpb++ = isc_tpb_no_rec_version;
-			dbh->transaction_flags &= ~(PDO_TRANS_ISOLATION_LEVEL^PDO_TRANS_READ_COMMITTED);
-		} else if (dbh->transaction_flags & PDO_TRANS_REPEATABLE_READ) {
-			*ptpb++ = isc_tpb_concurrency;
-			dbh->transaction_flags &= ~(PDO_TRANS_ISOLATION_LEVEL^PDO_TRANS_REPEATABLE_READ);
-		} else {
-			*ptpb++ = isc_tpb_consistency;
-			dbh->transaction_flags &= ~(PDO_TRANS_ISOLATION_LEVEL^PDO_TRANS_SERIALIZABLE);
-		}
-	}
 
-	if (dbh->transaction_flags & PDO_TRANS_ACCESS_MODE) {
-		if (dbh->transaction_flags & PDO_TRANS_READONLY) {
-			*ptpb++ = isc_tpb_read;
-			dbh->transaction_flags &= ~(PDO_TRANS_ACCESS_MODE^PDO_TRANS_READONLY);
-		} else {
-			*ptpb++ = isc_tpb_write;
-			dbh->transaction_flags &= ~(PDO_TRANS_ACCESS_MODE^PDO_TRANS_READWRITE);
+	/* isc_xxx are all 1 byte. */
+	char tpb[5] = { isc_tpb_version3 }, *ptpb = tpb + 1;
```

Comment:
If it is not an API with the FB_ prefix, it seems safe to assume that there is no difference between versions. Therefore, there are no differences between versions in the tpb specifications.

I think I can write it a little more clearly, so I'll try rewriting it.

Reply:
I tried to rewite it a little easier to understand.
https://github.com/php/php-src/pull/12815/commits/efa0d980bb095ffc08c00ed254a388edd61ed2a1

The setting add more 1 byte for `read committed`, because firebird's `read committed` supports multiple modes of behavior.

`isc_tpb_rec_version` is the mode for reading from committed data.
The opposite setting is `isc_tpb_no_rec_version`. This is a mode in which uncommitted data waits for it to be committed (or rolled back) and then read.

Detailed documentation on these:
https://docwiki.embarcadero.com/InterBase/2020/en/Specifying_Isolation_Level

Reply:
@Girgias 
Using strlen on char arrays doesn't seem to work well on windows. (It may be wrong to use strlen for arrays in the first place.)
I rewrote it for now and the test passed, but which one seems better, the one using pointers?